File: clang/test/CodeGenObjC/feature-availability.m
// RUN: %clang_cc1 -triple arm64-apple-macosx -fblocks -ffeature-availability=feature1:on -ffeature-availability=feature2:off -emit-llvm -o - %s | FileCheck %s

// CHECK: @"OBJC_CLASS_$_C0" =
// CHECK: @"OBJC_METACLASS_$_C0" =
// CHECK: @OBJC_CLASS_NAME_ = private unnamed_addr constant [3 x i8] c"C0\00",
// CHECK: @"_OBJC_METACLASS_RO_$_C0" =
// CHECK: @OBJC_METH_VAR_NAME_ = private unnamed_addr constant [3 x i8] c"m0\00",
// CHECK: @"_OBJC_$_INSTANCE_METHODS_C0" =
// CHECK: @"_OBJC_CLASS_RO_$_C0" =
// CHECK: @OBJC_CLASS_NAME_.1 = private unnamed_addr constant [5 x i8] c"Cat0\00",
// CHECK-NOT: C1
// CHECK-NOT: Cat1
// CHECK: @"OBJC_LABEL_CLASS_$" = private global [1 x ptr] [ptr @"OBJC_CLASS_$_C0"], section "__DATA,__objc_classlist,regular,no_dead_strip", align 8

__attribute__((feature(feature1)))
@interface C0
-(void)m0;
@end

// CHECK: define internal void @"\01-[C0 m0]"(

@implementation C0
-(void)m0 {
}
@end

@interface C0(Cat0)
@end

@implementation C0(Cat0)
@end

__attribute__((feature(feature2)))
@interface C1
-(void)m1;
@end

// CHECK-NOT: void @"\01-[C1 m1]"(
@implementation C1
-(void)m1 {
}
@end

@interface C1(Cat1)
@end

@implementation C1(Cat1)
@end

File: clang/test/CodeGen/feature-availability.c
// RUN: %clang_cc1 -triple arm64-apple-macosx -fblocks -ffeature-availability=feature1:on -ffeature-availability=feature2:off -emit-llvm -o - %s | FileCheck %s

// CHECK: %[[STRUCT_S0:.*]] = type { i32 }
// CHECK: @g0 = external global i32, align 4
// CHECK-NOT: @g1
// CHECK-NOT: @g2

__attribute__((feature(feature1))) int func0(void);
__attribute__((feature(feature2))) int func1(void);
int func2(void);

__attribute__((feature(feature1))) extern int g0;
__attribute__((feature(feature2))) int g1 = 100;
__attribute__((feature(feature2))) int g2;

struct __attribute__((feature(feature1))) S0 {
  int d0;
};

// CHECK-LABEL: define void @test0()
// CHECK-NOT: br
// CHECK: call i32 @func0()
// CHECK: store i32 123, ptr @g0, align 4
// CHECK-NOT: func1()
// CHECK-NOT: func2()
void test0(void) {
  if (__builtin_feature(feature1)) {
    func0();
    g0 = 123;
  }

  if (__builtin_feature(feature2)) {
    func1();
    g1 = 123;
  }

  if (__builtin_feature(feature1))
    if (__builtin_feature(feature2)) {
      func2();
    }
}

// CHECK-LABEL: define void @test1()
__attribute__((feature(feature1)))
void test1(void) {
}

// CHECK-NOT: @test2(
__attribute__((feature(feature2)))
void test2(void) {
}

// CHECK-LABEL: define void @test3(
// CHECK: %[[D0:.*]] = getelementptr inbounds nuw %[[STRUCT_S0]], ptr %{{.*}}, i32 0, i32 0
// CHECK: store i32 134, ptr %[[D0]], align 4
__attribute__((feature(feature1)))
void test3(struct S0 *s0) {
  s0->d0 = 134;
}
